lib/service/awsoidc.go

```diff
@@ -0,0 +1,301 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// cloudStableVersion has vX.Y.Z format, however the container image tag does not include the `v`.
+	cloudStableVersion := strings.TrimPrefix(stableVersion, "v")
+
+	databases, err := updater.AuthClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The updater needs to iterate over all integrations and aws regions to check
+	// for deploy service agents to update. In order to reduce the number of api
+	// calls, the aws regions are first reduced to only the regions containing
+	// an RDS database.
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	integrations, err := updater.AuthClient.ListAllIntegrations(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, ig := range integrations {
+		spec := ig.GetAWSOIDCIntegrationSpec()
+		if spec == nil {
+			continue
+		}
+		integrationName := ig.GetName()
+
+		for region := range awsRegions {
+			token, err := updater.AuthClient.GenerateAWSOIDCToken(ctx, types.GenerateAWSOIDCTokenRequest{
+				Issuer: updater.AWSOIDCProviderAddr,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			req := &awsoidc.AWSClientRequest{
+				IntegrationName: ig.GetName(),
+				Token:           token,
+				RoleARN:         spec.RoleARN,
+				Region:          region,
+			}
+
+			// The deploy service client is initialized using AWS OIDC integration
+			deployServiceClient, err := awsoidc.NewDeployServiceClient(ctx, req, updater.AuthClient)
+			if err != nil {
+				updater.Log.WithError(err).Warning("Failed to update deploy service agents.")
+				continue
+			}
+
+			ownershipTags := map[string]string{
+				types.ClusterLabel:     updater.TeleportClusterName,
+				types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+				types.IntegrationLabel: integrationName,
+			}
+
+			// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+			// If the lease cannot be acquired, the update is already being handled by another instance.
+			semLock, err := updater.AuthClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.SemaphoreKindConnection,
+				SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", region, integrationName),
+				MaxLeases:     1,
+				Expires:       updater.Clock.Now().Add(updateDeployAgentsInterval),
+				Holder:        "update_deploy_service_agents",
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), teleport.MaxLeases) {
+					updater.Log.Debug("Deploy service agent update is already being processed")
+					continue
+				}
+				return trace.Wrap(err)
+			}
+
+			updater.Log.Debugf("Updating Deploy Service Agents in AWS region: %s", region)
+			if err := awsoidc.UpdateDeployServiceAgent(ctx, deployServiceClient, awsoidc.UpdateServiceRequest{
```

Comment:
Given that we wait for the ECS service to be ready, I think we should parallelize per Integration/Region.
A waitgroup with maybe 3 slots should be ok (I don't think we should have unbounded parallelization)